symbolic-base/src/ZkFold/Symbolic/Data/MerkleTree.hs

```diff
@@ -0,0 +1,302 @@
+layerFolding :: forall c x.
+  ( SymbolicOutput x
+  , Context x ~ c
+  , SymbolicFold c
+  ) => List c x -> List c x
+layerFolding xs = res
+  where
+    (_, res) = foldr (Morph \(a :: Switch s x, (arr, l )) ->
+      ifThenElse (isNothing arr :: Bool s) (just a, l) (nothing, (newVer (fromJust arr) a) .: l))
+        (nothing :: Maybe c x, emptyList :: List c x) xs
+
+    -- (_, res) = foldr (Morph \(a :: Switch s x, (b :: Maybe s (Switch s x), l )) ->
+    --     (ifThenElse (isNothing b :: Bool s) (just a) nothing,
+    --      ifThenElse (isNothing b :: Bool s) l ((fromJust b) .: l)))
+    --                  (nothing :: Maybe c x, emptyList :: List c x) xs
+
+newVer ::
+  forall s y. (Symbolic s, SymbolicData y)
+  => Switch s y -> Switch s y -> Switch s y
+newVer l' r' = Switch (hashAux @s @y z (sLayout l') (sLayout r')) (sPayload l')
+  where
+    Bool z = false :: Bool s
+
+
+zeroMerkleTree :: forall d x c.
+  ( SymbolicFold c
+  , SymbolicOutput x
+  , Context x ~ c
+  , KnownNat d
+  , 1 <= d
+  , KnownNat (2 ^ (d-1))
+  , AdditiveMonoid x
+  ) => MerkleTree d x
+zeroMerkleTree = withDict (plusMinusInverse3 @1 @d) $ MerkleTree (bool (P.error "Invalid vector length") (arithmetize h Proxy) (L.null checkL)) (hl V..: ls)
+    where
+      (h, checkL) = L.uncons hl
+      (hl :: List c x, ls :: Vector (d-1) (List c x)) =
+        V.uncons @d @(List c x). V.reverse . Vector $ iterateN (P.fromIntegral $ value @d) layerFolding vs
+      vs :: List c x = P.foldr (L..:) emptyList $ fromVector (pure zero :: Vector (2 ^ (d-1)) x)
+
+instance forall c x d n.
+  ( SymbolicFold c
+  , SymbolicOutput x
+  , Context x ~ c
+  , KnownNat d
+  , 2 ^ (d-1) ~ n
+  , 1 <= d
+  ) => Iso (Vector n x) (MerkleTree d x) where
+  from v = withDict (plusMinusInverse3 @1 @d) $ MerkleTree (bool (P.error "Invalid vector length") (arithmetize h Proxy) (L.null checkL)) (hl V..: ls)
+    where
+      (h, checkL) = L.uncons hl
+      (hl :: List c x, ls :: Vector (d-1) (List c x)) =
+        V.uncons @d @(List c x). V.reverse . Vector $ iterateN (P.fromIntegral $ value @d -! 1) layerFolding vs
+      vs :: List c x = P.foldr (L..:) emptyList $ fromVector @n v
+
+instance forall c x d n.
+  ( SymbolicFold c
+  , SymbolicOutput x
+  , Context x ~ c
+  , KnownNat d
+  , 2 ^ (d-1) ~ n
+  , 1 <= d
+  ) => Iso (MerkleTree d x) (Vector n x)  where
+  from (MerkleTree _ l) = V.unsafeToVector @n $ helper @c (V.last l) (2 ^ (value @d -! 1))
+    where
+      helper :: forall s y. (SymbolicOutput y, Symbolic s, Context y ~ s) => List s y -> Natural -> [y]
+      helper ls k = case k of
+        0 -> []
+        _ -> let (n, ns) = L.uncons ls in n : helper @s ns (k-!1)
+
+hashAux :: forall c x.
+  ( Symbolic c
+  , SymbolicData x)
+  => c Par1 -> c (Layout x) -> c (Layout x) -> c (Layout x)
+hashAux b' h g = do
+  let (Vec v1) = merkleHasher [Vec h, Vec g]
+      (Vec v2) = merkleHasher [Vec g, Vec h]
+  fromCircuit3F b' v1 v2 $ \ (Par1 b) x y ->
```

Comment:
Here, an `ifThenElse` can be used from `instance Conditional (Bool c) (Vec f c)`